pep-0707.rst

```diff
@@ -0,0 +1,380 @@
+case of ``__exit__``, and with this to initiate a discussion of our options
+for evolving the signature of ``__exit__``.
+
+In the case of the :mod:`py3.11:traceback` module's API, evolving the
+functions to have a hybrid signature is relatively straighforward and
+safe. The functions take one positional and two optional arguments, and
+interpret them according to their types. This is safe when sentinels
+are used for default values.  The signatures of callbacks, which are
+defined by the user's program, are harder to evolve.
+
+The safest option is to make the user explicitly indicate which signature
+the callback is excpecting, by marking it with an additional attribute or
+giving it a different name. For example, we could make the interpreter
+look for a ``__leave__`` method on the context manager, and call it with
+a single arg if it exists (otherwise, it looks for ``__exit__`` and
+continues as it does now). The introspection-based alternative proposed
+here intends to make it more convenient for users to write new code,
+because they can just use the single-arg version and remain unaware of
+the legacy API. However, if the limitations of introspection are found
+to be too severe, we should consider an explicit option. Having both
+``__exit__`` and ``__leave__`` around for 5-10 years with similar
+function is not ideal, but it is an option.
+
+Let us now examine the limitations of the current proposal. It identifies
+2-arg python functions, and ``METH_O`` C functions as having a single-arg
+signature, and assumes that anything else is expecting 3 args. Obviously
+it is possible to create false negatives for this heuristic (single-arg
+callables that it will not identify). Context managers written in this
+way won't work, they will continue to fail as they do now when their
+``__exit__`` function will be called with three arguments.
+
+I believe that it will not be a problem in practice. First, all working
+code will continue to work, so this is a limitation on new code rather
+than a problem impacting existing code. Second, exotic callable types are
+rarely used for ``__exit__`` and if one is needed, it can always be wrapped
+by a plain vanilla method that delegates to the callable. For example, we
+can write this::
+
+   class C:
+      __enter__ = lambda self: self
+      __exit__ = ExoticCallable()
+
+as follows::
+
+   class CM:
+      __enter__ = lambda self: self
+      _exit = ExoticCallable()
+      __exit__ = lambda self, exc : CM._exit(exc)
+
+
+While discussing the real-world impact of the problem of this PEP, it is
+worth noting that most ``__exit__`` functions don't do anything with their
+arguments. Typically, a context manager is implemented to ensure that some
+cleanup actions take place upon exit. It is rarely appropriate for the
+``__exit__`` function to handle exceptions raised within the context, and
+they are typically allowed to propagate out of ``__exit__`` to the calling
+function.
```

Comment:
Maybe make a more explicit mention of how this actually impacts both the need for this PEP but also the backward-compat concern?